File: Backend/controllers/camerasController.js
/**
 * Cameras Controller
 * Manage additional cameras per user
 */

const User = require('../models/User');
const { generateRtspUrlFromCamera } = require('./authController');

/**
 * @desc    Get all cameras for current user (including primary as first entry)
 * @route   GET /api/cameras
 * @access  Private
 */
const getCameras = async (req, res, next) => {
  try {
    const user = await User.findById(req.user._id);
    if (!user) {
      return res.status(404).json({ success: false, error: 'User not found' });
    }

    const primaryCamera = {
      id: 'primary',
      name: user.cctvName,
      rtspUrl: user.rtspUrl,
      location: user.location,
      brand: undefined,
    };

    const extraCameras = (user.cameras || []).map((cam, index) => ({
      id: cam._id?.toString() || `extra-${index}`,
      name: cam.name,
      rtspUrl: cam.rtspUrl,
      location: cam.location,
      brand: cam.brand,
    }));

    res.json({
      success: true,
      data: [primaryCamera, ...extraCameras],
    });
  } catch (error) {
    next(error);
  }
};

/**
 * @desc    Add a new camera for current user
 * @route   POST /api/cameras
 * @access  Private
 */
const addCamera = async (req, res, next) => {
  try {
    const { name, location, rtspUrl, cameraIp, cameraUsername, cameraPassword, cameraPort, cameraBrand, cameraPath } = req.body;

    if (!name || !location) {
      return res.status(400).json({ success: false, error: 'Camera name and location are required' });
    }

    let finalRtspUrl = rtspUrl;

    if (!finalRtspUrl) {
      finalRtspUrl = generateRtspUrlFromCamera({
        ip: cameraIp,
        username: cameraUsername,
        password: cameraPassword,
        port: cameraPort,
        brand: cameraBrand,
        path: cameraPath,
      });
    }

    if (!finalRtspUrl) {
      return res.status(400).json({ success: false, error: 'Unable to generate RTSP URL from provided camera details' });
    }

    const user = await User.findById(req.user._id);
    if (!user) {
      return res.status(404).json({ success: false, error: 'User not found' });
    }

    user.cameras = user.cameras || [];
    user.cameras.push({
      name,
      rtspUrl: finalRtspUrl,
      location,
      brand: cameraBrand,
    });

    await user.save();

    res.status(201).json({ success: true, message: 'Camera added successfully' });
  } catch (error) {
    next(error);
  }
};

/**
 * @desc    Delete a camera for current user
 * @route   DELETE /api/cameras/:id
 * @access  Private
 */
const deleteCamera = async (req, res, next) => {
  try {
    const { id } = req.params;
    const user = await User.findById(req.user._id);
    if (!user) {
      return res.status(404).json({ success: false, error: 'User not found' });
    }

    if (id === 'primary') {
      // Clear primary camera fields
      user.rtspUrl = '';
      user.cctvName = '';
      await user.save();
      return res.json({ success: true, message: 'Primary camera removed' });
    }

    // Remove from cameras array
    const camIndex = (user.cameras || []).findIndex(
      (cam) => cam._id?.toString() === id
    );
    if (camIndex === -1) {
      return res.status(404).json({ success: false, error: 'Camera not found' });
    }

    user.cameras.splice(camIndex, 1);
    await user.save();

    res.json({ success: true, message: 'Camera removed successfully' });
  } catch (error) {
    next(error);
  }
};

module.exports = {
  getCameras,
  addCamera,
  deleteCamera,
};

File: Backend.NET/src/AttackShield.API/Controllers/CamerasController.cs
using AttackShield.Core.DTOs;
using AttackShield.Core.Entities;
using AttackShield.Core.Interfaces;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace AttackShield.Api.Controllers;

[Authorize]
public sealed class CamerasController : ApiControllerBase
{
    private readonly IUserRepository _users;
    private readonly IRtspUrlBuilder _rtsp;
    private readonly IStreamManager _streams;

    public CamerasController(IUserRepository users, IRtspUrlBuilder rtsp, IStreamManager streams)
    {
        _users = users;
        _rtsp = rtsp;
        _streams = streams;
    }

    [HttpGet]
    public async Task<IActionResult> Get(CancellationToken ct)
    {
        var user = await GetUser(ct);
        if (user is null) return Fail("User not found", 404);

        var cameras = new List<CameraDto>();
        if (!string.IsNullOrWhiteSpace(user.RtspUrl))
            cameras.Add(new("primary", user.CctvName, user.RtspUrl, user.Location, null));
        cameras.AddRange(user.Cameras.Select((camera, index) =>
            new CameraDto(camera.Id ?? $"extra-{index}", camera.Name, camera.RtspUrl, camera.Location, camera.Brand)));
        return Ok(new { success = true, data = cameras });
    }

    [HttpPost]
    public async Task<IActionResult> Add([FromBody] AddCameraRequest request, CancellationToken ct)
    {
        if (string.IsNullOrWhiteSpace(request.Name) || string.IsNullOrWhiteSpace(request.Location))
            return Fail("Camera name and location are required");

        var rtspUrl = request.RtspUrl ?? _rtsp.Build(request.CameraIp, request.CameraUsername,
            request.CameraPassword, request.CameraPort, request.CameraBrand, request.CameraPath);
        if (string.IsNullOrWhiteSpace(rtspUrl))
            return Fail("Unable to generate RTSP URL from provided camera details");

        var user = await GetUser(ct);
        if (user is null) return Fail("User not found", 404);
        user.Cameras.Add(new EmbeddedCamera
        {
            Id = MongoDB.Bson.ObjectId.GenerateNewId().ToString(),
            Name = request.Name,
            Location = request.Location,
            RtspUrl = rtspUrl,
            Brand = request.CameraBrand,
        });
        user.UpdatedAt = DateTime.UtcNow;
        await _users.UpdateAsync(user.Id!, user, ct);
        return StatusCode(201, new { success = true, message = "Camera added successfully" });
    }

    [HttpDelete("{id}")]
    public async Task<IActionResult> Delete(string id, CancellationToken ct)
    {
        var user = await GetUser(ct);
        if (user is null) return Fail("User not found", 404);

        if (id == "primary")
        {
            user.RtspUrl = string.Empty;
            user.CctvName = string.Empty;
            user.UpdatedAt = DateTime.UtcNow;
            await _users.UpdateAsync(user.Id!, user, ct);
            await _streams.StopAsync($"{user.Id}-primary", ct);
            return Ok(new { success = true, message = "Primary camera removed" });
        }

        var camera = user.Cameras.FirstOrDefault(c => c.Id == id);
        if (camera is null) return Fail("Camera not found", 404);
        user.Cameras.Remove(camera);
        user.UpdatedAt = DateTime.UtcNow;
        await _users.UpdateAsync(user.Id!, user, ct);
        await _streams.StopAsync($"{user.Id}-{id}", ct);
        return Ok(new { success = true, message = "Camera removed successfully" });
    }

    private Task<User?> GetUser(CancellationToken ct)
        => CurrentUserId is null ? Task.FromResult<User?>(null) : _users.GetByIdAsync(CurrentUserId, ct);
}
